Custom workloads

NathanTP opened the pull request:
This is a major overhaul of the workload generation system. It introduces custom workloads with a real build system and launch/testing framework. The first release will not introduce significant new workloads (all existing workloads still live in firesim/deploy/workloads), the porting effort will happen incrementally.

See the corresponding PR in firesim.git for full documentation.

sagark commented:
@NathanTP can you fix the merge conflicts against dev here?

NathanTP commented:
I merged dev this should fix any merge conflicts. This branch changes too many things for a real merge though, so I used the 'ours' strategy (ignore the existing branch).

davidbiancolin commented on README.md:
```diff
@@ -1,10 +1,14 @@
-firesim-software
+Fire Marshal (firesim-software)
```
I propose we reduce this readme to a few lines that aggressively redirects to FireSim's docs.

NathanTP replied:
Agreed. Fixed.

davidbiancolin commented on full_test.sh:
```diff
@@ -0,0 +1,85 @@
+SUITE_PASS=true
+
+# Reset the test log
+echo "" > test.log
```
Why not just a new file?

NathanTP replied:
fixed. It now uses mktemp to create a random file name for each run and reports that to the user.

davidbiancolin commented on test/incremental/README.md:
```diff
@@ -0,0 +1,23 @@
+Tests the incremental building capabilities of the system. The test procedure
+is fairly specific, so it's best to always use the test.py script to test this. 
+
+Incremental builds means that sw\_manager won't rebuild images from the base
```
sw_manager -> marshal

NathanTP replied:
fixed

davidbiancolin commented on wlutil/br/br.py:
```diff
@@ -0,0 +1,107 @@
+    def upToDate(self):
+        # XXX There's something wrong with buildroots makefile, it throws an
+        # error and never reports being up to date.
+        # XXX DONT COMMIT THIS CHANGE YOUR DEFNITELY GOING TO FORGET TO UNDO THIS
```
This needs to be dealt with.

NathanTP replied:
That comment was from when I had it hard-coded to never even check. Now it checks if there is a buildroot output, if not, it will rebuild. If there is some output, it will use it. This means it won't rebuild buildroot if you change some core thing in the buildroot distribution (like buildroot-config) or the buildroot overlay. To change those things, just run make in wlutil/br. Without this, every buildroot-based workload gets fully reproduced every time (10x+ build time).

The comment has been fixed, and I've added a readme to wlutil/br to clarify the situation. In practice, this isn't really an issue because even init scripts and stuff are handled through firemarshal overlays instead of buildroot overlays. We could (and maybe should at some point) replace buildroot with a pre-built image like we do for Fedora.

NathanTP commented:
I cleaned up some unnecessary complexity in tests and examples. We're down to 8k new lines instead of 26k :).

I think I've addressed all the open issues.

davidbiancolin commented on wlutil/launch.py:
```diff
@@ -0,0 +1,96 @@
+from .wlutil import *
+
+# The amount of memory to use when launching
+launch_mem = "16384"
```
This smells a little. Is this exposed somewhere?

NathanTP replied:
This just matches the default firesim config. It's not exposed right now. I think stuff like this needs to go into a "board" abstraction for FireMarshal. I also think that the 'launch' capability may merge with FireSim at some point which would also require some new abstractions (a virtual agfi or something).
